### PR TITLE
Started LlpcOverview.md and LgcOverview.md documents

### PR DESCRIPTION
**Pull request by trenouf**
Part of a series of patches to clean up the middle-end interface.

These new files outline the state of LLPC and LGC (the middle-end part
of LLPC) as they will be at the end of the "middle-end interface
clean-up" series of patches.

LgcOverview.md contains the material that was in the big comment at the
top of Builder.h.

Change-Id: Ie1b21f5ddbba597c80c5d833d8abe0c9685a6c36

### CODE REVIEW

File: docs/LgcOverview.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+The front-end flow is:
+
+* Call `BuilderContext::Initialize`. The first time this is called must be before LLVM command-line
+  processing, and before the first call to `BuilderContext::Create`. It is allowed to call it multiple
+  times.
+
+* Process command-line options with `cl::ParseCommandLineOptions`. This can only be done once in the
+  whole process.
+
+* Create a BuilderContext using `BuilderContext::Create`, passing in
```
Comment (linqun) [inline docs/LgcOverview.md]: It seems this section isn't well formatted, each paragraph has a *, is it expected? AsciidocFX is used in my local machine.

File: docs/LlpcOverview.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+"middle-end interface clean-up" series of patches.
+
+## Compilation process
+
+The LLPC compilation process consists of three phases:
+* front-end;
+* middle-end (LGC);
+* back-end (LLVM AMDGPU back-end).
+
+### Front-end
```
Comment (linqun) [inline docs/LlpcOverview.md]: I don't why this tag isn't shown correctly in my local machine, could you double check it. thanks!

File: docs/LlpcOverview.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+See
+[LGC (LLPC middle-end) overview](LgcOverview.md)
+for more details.
+
+* In BuilderRecorder mode, run the BuilderReplayer pass to lower the `llpc.call.*` calls
+  into actual IR, and read the pipeline state metadata.
+* Perform further lowering on the LLVM IR to obtain LLVM IR with `llvm.amdgcn.*` intrinsics
+  understood by the AMDGPU back-end, and the PAL metadata needed by the back-end to
+  add to the ELF.
+* Run middle-end optimizations, most of which are standard LLVM ones.
```
Comment (linqun) [inline docs/LlpcOverview.md]: Could you high-light some special works in LGC, include NGG, merge shader, assign shader register value and etc in some place, either in this doc or in LgcOverview.md. thanks!

File: docs/LgcOverview.md (hunk 1 above)
Comment (trenouf) [reply docs/LgcOverview.md]: You can see the formatted markdown in github by browsing the commit and selecting the md file. Github then shows it with formatting. E.g.
https://github.com/trenouf/llpc/blob/pr380-overview/docs/LgcOverview.md

Using that view, the formatting looks ok to me.

File: docs/LlpcOverview.md (hunk 2 above)
Comment (trenouf) [reply docs/LlpcOverview.md]: As above; looks ok to me when browsing in github.

File: docs/LlpcOverview.md (hunk 3 above)
Comment (trenouf) [reply docs/LlpcOverview.md]: Good idea, I will do.

File: docs/LgcOverview.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+* The second kind of outside-IR state is the state set in middle-end passes (and BuilderImpl).
+  This is `ResourceUsage` and `InterfaceData`, which are declared in their own separate file
+  `llpcResourceUsage.h`.
+  This state does not get written into IR metadata. In theory we could add a facility to write
+  the state into IR metadata and read it back, and that would allow us to support lit testing of
+  individual middle-end passes with `-stop-before` and `-run-pass` options like `llc`. But
+  that does not happen now.
+
+### ShaderModes
```
Comment (amdrexu) [inline docs/LgcOverview.md]: I don't understand what ShaderModes represent? Which info will it record?

File: docs/LlpcOverview.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
+# LLPC overview
+
+This document outlines the state of LLPC as it will be at the end of the
+"middle-end interface clean-up" series of patches.
+
+## Compilation process
+
+The LLPC compilation process consists of three phases:
+* front-end;
```
Comment (amdrexu) [inline docs/LlpcOverview.md]: Is it possible to add a data flow diagram to show the compilation or the top-level  constructs. I mean just like this

```
+----+
| FE |  ----->
+----+
```

File: docs/LlpcOverview.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+* All the usual back-end transformations (instruction selection, control-flow lowering
+  for SIMD, scheduling, register allocation, ISA emission).
+
+The result is an ELF that is PAL ABI compliant.
+
+## Pipeline and shader caching
+
+TODO: Outline of how the rest of pipeline and shader caching works in the front-end.
+
+### Part-pipeline caching
```
Comment (amdrexu) [inline docs/LlpcOverview.md]: Part -> Partial

File: docs/LgcOverview.md (hunk 1 above)
Comment (linqun) [reply docs/LgcOverview.md]: got it. please ignore my previous comments. thanks!

**Comment by trenouf** [comment]
V5: Addressed review comments, including adding a diagram.

File: docs/LgcOverview.md (hunk 4 above)
Comment (trenouf) [reply docs/LgcOverview.md]: Added more info there.

File: docs/LlpcOverview.md (hunk 5 above)
Comment (trenouf) [reply docs/LlpcOverview.md]: I made the diagram a bit more detailed than that. :-)

**Comment by trenouf** [comment]
V6: Slight diagram adjustment

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9ec3c84117a611ec5ca5b5f8d6f3fb629fbc8f5f
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 831f334a9f4ef429b5d4e953b86b4624a7c2db54</li>
<li>SPVGEN: dev f1bc2ba988273c3724afffe72fe9cd933a022ce7</li>
<li>XGL: dev 54a1ba508fd82ed2a11d0960f0f2f35f975b2f12</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev cc0df5ace776584f5f7c0c20704d28f445f0e074</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 939/939 (100.0%)</li>
<li>Failed: 0/939 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 65433/65433 (100.0%)</li>
<li>Failed: 0/65433 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 65465/65465 (100.0%)</li>
<li>Failed: 0/65465 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 65465/65465 (100.0%)</li>
<li>Failed: 0/65465 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**